### DISCUSSION TITLE
Why not Svelte?

### DISCUSSION THREAD

**Discussion by frederikhors** [chakra-ui/zag#262] [Q&A] [answered]
# 🚀 Feature request

Vue, React, Solid, why not Svelte?

**Comment by segunadebayo** [comment]
Hi @frederikhors,

The only thing holding us back from supporting Svelte is https://github.com/sveltejs/svelte/issues/5112

An important aspect of Zag is that we manage the event handlers internally and provide a way for users to spread it onto the elements. Svelte doesn't support events in spread props just yet.

When it does, we'll add support for it.

**Comment by frederikhors** [comment]
Can we leave this open?

**Comment by anubra266** [comment]
> Can we leave this open?

Yeah sure, It'll help others looking for Svelte understand why too.

**Comment by impactvelocity** [comment]
Looks like some progress for svelte is coming? :)

https://github.com/chakra-ui/zag/tree/main/packages/frameworks/svelte

I tried it out, but couldn't figure out the api.connect normalizeProps.

**Comment by csenio** [comment]
I managed to get it working by making use of svelte actions!! https://codesandbox.io/s/zag-svelte-n01f02?file=/Select.svelte

@segunadebayo Is there a scenario where that wouldn't work?

**Reply by anubra266** [reply]
Wow...this seems promising @jescowuester

**Reply by csenio** [reply]
This library is incredible. Did a more advanced example with the dialog component and everything **just works**: https://codesandbox.io/s/zag-svelte-menu-4kyv8q?file=/App.svelte

**Reply by anubra266** [reply]
@jescowuester I opened a draft PR [here](https://github.com/chakra-ui/zag/pull/550) where, we can fight all the nuances.
1. I tried accordion, and there seems to be a render issue where the items all flash on initial render, very possible such will happen for other components.
2. The sample uses sveltekit, there seems to be a problem with initial render, where zag is trying to access the browser before it's available, so it throws a 500
...And whatever else we come across. Really happy that you brought up this method.
PS: I'm not badass at Svelte. 😁

**Reply by anubra266** [reply]
Confirmed @jescowuester. SSR is the drawback on using actions.

**Reply by TGlide** [reply]
I've seen some movement on the `svelte-components` branch. Does this mean zagjs for svelte is underway?

**Reply by anubra266** [reply]
Yeah @TGlide we're trying to make svelte work. We're only facing one issue. When inputs value are set programmatically in other frameworks, `onChange` is called, svelte doesn't do this. We're looking for the best way to patch it up.

**Reply by TGlide** [reply]
> Yeah @TGlide we're trying to make svelte work. We're only facing one issue. When inputs value are set programmatically in other frameworks, `onChange` is called, svelte doesn't do this. We're looking for the best way to patch it up.

I see! Could you maybe make the action dispatch a custom event on change?

EDIT: This is probably too naive and you probably thought about it. I'll try around messing with the code though and see if I can get anywhere. thanks!

**Reply by anubra266** [reply]
@TGlide mentioning custom events actually gave me an idea. Dispatching a custom event in the agnostic logic, and listening for it in the svelte adapter.

**Reply by TGlide** [reply]
> @TGlide mentioning custom events actually gave me an idea. Dispatching a custom event in the agnostic logic, and listening for it in the svelte adapter.

Oh nice! I'd love to know how that goes 🤩

**Reply by TGlide** [reply]
Also, I have a (hacky) way to pass in event listeners without actions: https://svelte.dev/repl/daf364f729e54818856d7e718c208fb9?version=3.59.1

I did some experiments over on my library, if you're curious to see a more advanced implementation:
- Store with event listeners: https://github.com/TGlide/radix-svelte/blob/feat/builders/src/lib/internal/stores/derived.ts
- Component using said store: https://github.com/TGlide/radix-svelte/blob/feat/builders/src/lib/builders/collapsible/index.ts
- Component usage: https://github.com/TGlide/radix-svelte/blob/feat/builders/src/routes/builders/collapsible/preview.svelte

Don't know if this is something that could/should be used in Zag, but thought it'd be nice to share :)

**Reply by anubra266** [reply]
@TGlide we don't use actions for attributes, just events..
And we can't keep logic in stores because Zag is agnostic.

**Reply by TGlide** [reply]
I meant events, that was a typo on my end. And I see, makes sense!

**Comment by Not-Jayden** [comment]
@segunadebayo  FYI Svelte 5 is moving away from `on:x` syntax to `onx` event props that can be spread directly onto elements ([example](https://svelte-5-preview.vercel.app/#H4sIAAAAAAAAE42PQUvEMBCF_8oQhO3ikuq1tgvr37Ae1uwUgs0kJBNBQv67SbOuCB485DAz7718L4lFrxjE8JIEnQ2KQZycEwfBn64O4QNXxjIHG72qmzEorx0fZ5pZG2c9w3NktgSLtwZ2sm-jbNbdUxXWtyKDspEYJrgLfGbsHvb1Wm5LJMW6ZGhSHg0Sd3tIm5Ob536Cxy0qzzT2Pww0Xn9Pmy6DJbVq9T6lW1SG_lgaGHvRi8aLGNhHzIdb4Rbw3861RpJSegwluVRx3rrQ1Sa_ud6uXN_azb2xhQFSXcmGXI1N_Afma_4CTTYp06MBAAA=))

Would love to put my hand up to work on this (and subsequently the Ark UI implementation). Svelte 5 is still in alpha but keen to get a head start.

**Reply by anubra266** [reply]
@Not-Jayden the other issue with svelte is events handling.

> The main blocking point was the way the dispatch method works. React/Vue/Solid etc, do stuff behind the scenes that changes the native way of how events are triggered programmatically, to make it more natural, while Svelte stays very close to the native way. 
We were very close to supporting svelte in Zag but this was a huge blocking point since it meant some component worked differently in Svelte and  required us to build this kind of stuff for Svelte ( which is possible, just not maintainable )

Examples of what other frameworks do on some events :
React: https://github.com/facebook/react/blob/main/packages/react-dom-bindings/src/events/plugins/ChangeEventPlugin.js#L281
Vue and Preact also do this.

**Reply by cschroeter** [reply]
Another aspect to consider for Ark UI is the support of the `asChild` prop. While it's not mandatory, lacking support for this could lead to inconsistencies. It also raises questions about how to effectively merge logical components like a `DialogTrigger` with the users `Button` component

**Reply by Not-Jayden** [reply]
Maybe I'll explore this for myself in a separate repo and see where I get. I don't fully understand what the limitations are yet with achieving a spec-compatible implementation, but I'm sure it can get at least 95% of the way there.

**Reply by cschroeter** [reply]
@Not-Jayden 

I'm happy to help.

**Reply by ditorodev** [reply]
Im curious if there has been progress here? This is very promising, if theres something I can help with im happy to (sporadically, don't have a bunch of time)

**Reply by Not-Jayden** [reply]
Not from me personally. I was going to re-assess once Svelte 5 reaches general availability as there's been a fair few significant changes since I first commented. I imagine that should be within a couple months or so.

If anyone is super keen to get a jump on it now though please feel free though.

**Reply by gbouteiller** [reply]
@Not-Jayden actually, i'm currently working on it as we speak. I'm working in parallel on the docs and i'll have the beginning of something next week.

**Comment by segunadebayo** [comment, marked as answer]
We've got the first installment of Svelte support merged. Although not publicly available yet, we'll polish it a bit more and prepare it for release pretty soon.

**Reply by frederikhors** [reply]
Any news?

**Reply by anubra266** [reply]
@frederikhors  Most things are in place already; you can see try it, snippets are [here](https://github.com/chakra-ui/zag/tree/main/website/data/snippets/svelte).
Basically just waiting for a stable svelte v5, then we can add test suite and go live.

**Reply by HummingMind** [reply]
> @frederikhors Most things are in place already; you can see try it, snippets are [here](https://github.com/chakra-ui/zag/tree/main/website/data/snippets/svelte). Basically just waiting for a stable svelte v5, then we can add test suite and go live.

Svelte 5 is out. 🥳
